src/v/raft/consensus.cc

```diff
@@ -2389,74 +2390,77 @@ void consensus::maybe_update_leader_commit_idx() {
         vlog(_ctxlog.warn, "Error updating leader commit index", e);
     });
 }
-
+/**
+ * The `maybe_commit_configuration` method is the place where configuration
+ * state transition is decided and executed
+ */
 ss::future<> consensus::maybe_commit_configuration(ss::semaphore_units<> u) {
     // we are not a leader, do nothing
     if (_vstate != vote_state::leader) {
-        return ss::now();
+        co_return;
     }
 
     auto latest_offset = _configuration_manager.get_latest_offset();
     // no configurations were committed
     if (latest_offset > _commit_index) {
-        return ss::now();
+        co_return;
     }
 
     auto latest_cfg = _configuration_manager.get_latest();
+
     /**
-     * simple configuration, there is nothing to do
+     * current config still contains learners, do nothing
      */
-    if (latest_cfg.type() == configuration_type::simple) {
-        return ss::now();
+    if (unlikely(!latest_cfg.current_config().learners.empty())) {
+        co_return;
     }
-    /**
-     * at this point joint configuration is committed, and all added learners
-     * were promoted to voter
-     */
-    if (latest_cfg.current_config().learners.empty()) {
-        /*
-         * In the first step we demote all removed voters to learners
-         */
-
+    switch (latest_cfg.type()) {
```

Comment:
Not sure if what I am going to suggest is reasonable or easy to implement, but anyway I'd like to discuss it.

Here we have the reconfiguration state machine and this state machine is driven internally - all needed state is contained in the `group_configuration` struct and here after the configuration is committed we figure out the next step based on that state. This has some unfortunate side effects:
1) we have to duplicate information about added/removed nodes (that we already have in the controller log) and serialize it
2) state machine transitions become coupled to the serialization format. I.e. if we would want to change the transition rules, it probably would mean introducing the new version and figuring out how to represent the needed state in the `group_configuration` struct (just like now we use the presence of the `_configuration_update` field to deduce that we are in the first phase of the transition).

We also now have cancellation - which means that this state machine is now at least partially externally driven. So my question is - can we make it fully externally driven (or at least delegate calculating the transitions to the higher level)? Configuration will then become a simple stateless struct describing the current configuration of the raft group and all the complicated logic related to configuration transitions will be at the level of `controller_backend` where all additional needed info is readily available.

Reply:
I agree that we may refactor the code to extract configuration change logic to some  higher level abstraction layer. I think it shouldn't be `controller_backend` tho. The whole configuration change driving state machine must be part of raft as it must operate under raft operation mutex. This is why i decided on adding information about nodes being added/removed to configuration structure itself. 

We may plan to refactor raft configuration handling to decouple configuration data structure from the logic driving the configuration changes. (i haven't decided to do it here as the code churn would be pretty large)